scripts/hud/customizer.ts

```diff
@@ -0,0 +1,654 @@
+
+		if (this.dragMode === DragMode.MOVE) {
+			panelPos = LayoutUtil.getPosition(this.panels.virtual);
+		} else {
+			const resizeDir = resizeVector.get(this.dragMode) ?? [0, 0];
+			const knobPos = LayoutUtil.getPosition(this.panels.virtualKnob);
+			for (const i of Axes) {
+				if (resizeDir[i] === 1) {
+					panelSize[i] = Math.max(knobPos[i] - panelPos[i], minSize[i]);
+				} else if (resizeDir[i] === -1) {
+					panelSize[i] += panelPos[i] - knobPos[i];
+					const offset = Math.max(panelSize[i], minSize[i]) - panelSize[i];
+					panelSize[i] += offset;
+					panelPos[i] = knobPos[i] - offset;
+				}
+			}
+		}
+
+		this.updateVirtualPanelFontSize();
+
+		LayoutUtil.setPositionAndSize(this.panels.virtual, panelPos, panelSize);
+
+		// snapping
+		if (this.dragMode === DragMode.MOVE) {
+			for (const axis of Axes) {
+				const isX = axis === 0;
+
+				if (this.activeComponent.snaps[axis] !== SnapMode.OFF) {
+					const sizeFactor = snaps[axis][this.activeComponent.snaps[axis]].sizeFactor;
+
+					const offset = panelSize[axis] * sizeFactor;
+
+					const gridline = this.getNearestGridLine(axis, sizeFactor);
+					const activeGridline = this.activeGridlines[axis];
+
+					if (gridline) panelPos[axis] = gridline.offset - offset;
+					if (gridline !== activeGridline) {
+						if (activeGridline) activeGridline.panel.RemoveClass('hud-customizer__gridline--highlight');
+						if (gridline) {
+							gridline.panel.AddClass('hud-customizer__gridline--highlight');
+							this.activeGridlines[axis] = gridline;
+						}
+					}
+				}
+			}
+		}
+
+		this.activeComponent.position = panelPos;
+		this.activeComponent.size = panelSize;
+		LayoutUtil.setPositionAndSize(
+			this.activeComponent.panel,
+			this.activeComponent.position,
+			this.activeComponent.size
+		);
+		this.updateResizeKnobs(panelPos, panelSize);
+	}
+
+	private onEndDrag() {
+		if (!this.activeComponent) return;
+
+		this.onDragThink(); // call the last think to ensure the final position is set
+
+		$.UnregisterEventHandler('HudThink', $.GetContextPanel(), this.onThinkHandle);
+
+		LayoutUtil.setPositionAndSize(this.panels.virtual, this.activeComponent.position, this.activeComponent.size);
+
+		this.activeComponent.panel.RemoveClass('hud-customizable--dragging');
+		this.panels.virtual.RemoveClass('hud-customizer-virtual--dragging');
+
+		this.activeGridlines?.forEach((line) => line?.panel.RemoveClass('hud-customizer__gridline--highlight'));
+		this.activeGridlines = [undefined, undefined];
+
+		// TODO: this is just for testing
+		this.save();
+
+		this.dragMode = undefined;
+	}
+
+	private updateResizeKnobs(position: LayoutUtil.Position, size: LayoutUtil.Size): void {
+		const [width, height] = size;
+		const halfWidth = width / 2;
+		const halfHeight = height / 2;
+		let deltaX: number, deltaY: number;
+		for (const [dir, knob] of Object.entries(
+			this.resizeKnobs satisfies Record<ResizeDirection, Button>
+		) as unknown as [ResizeDirection, Button][]) {
```

Comment:
TS is making you do this ugly casts because since Object keys are always strings, `dir` isn't really a `ResizeDirection`, it's a string.
```ts
for (const [dir, knob] of Object.entries(this.resizeKnobs) {
  switch (+dir as ResizeDirection) {
    ...
```
I think TS will still be able to infer that `knob` is a `Button` in that case.

I'm always on the fence about objects vs maps for structures like this, maps not stringifying their keys is nice but syntax is quite a lot uglier, especially for initialization. Could convert this to a map if we wanted, but if so we should probably be consistent and convert `snaps` as well.

Reply:
Addressed in 74f396aa55355f1288479bab2290b85024d24fa6 (in a more type-safe way, but can change back to just `+dir as ResizeMode` for readability)

Reply:
I checked out your branch to test this out and TypeScript appears to be fine with 
```ts
for (const [dir, knob] of Object.entries(this.resizeKnobs)) {
  switch (+dir) {
```
where `this.resizeKnobs` is just `private resizeKnobs: Record<ResizeMode, Button>;`.

IMO that approach is completely fine, TypeScript did the important thing in this case, flagging that switching over `dir` with numeric values of `DragMode` must be comparing a string to a number. Whilst I personally enjoy doing arcane type manipulation to prove stuff to the compiler it tends to lead to code smells and worse maintainability. The template literal stuff was really neat though!

Reply:
Addressed in b5c6a791f3d7e42dd3c4bc280c8a0663aadbf9fc